lib/Dialect/Arc/Transforms/LowerState.cpp

```diff
@@ -0,0 +1,593 @@
+/// non-null operand, or null if both operands are null.
+Value ClockLowering::getOrCreateAnd(Value lhs, Value rhs, Location loc) {
+  if (!lhs)
+    return rhs;
+  if (!rhs)
+    return lhs;
+  auto &slot = andCache[std::make_pair(lhs, rhs)];
+  if (!slot)
+    slot = builder.create<comb::AndOp>(loc, lhs, rhs);
+  return slot;
+}
+
+//===----------------------------------------------------------------------===//
+// Module Lowering
+//===----------------------------------------------------------------------===//
+
+GatedClockLowering ModuleLowering::getOrCreateClockLowering(Value clock) {
+  // Look through clock gates.
+  if (auto ckgOp = clock.getDefiningOp<ClockGateOp>()) {
+    // Reuse the existing lowering for this clock gate if possible.
+    if (auto it = gatedClockLowerings.find(clock);
+        it != gatedClockLowerings.end())
+      return it->second;
+
+    // Get the lowering for the parent clock gate's input clock. This will give
+    // us the clock tree to emit things into, alongside the compound enable
+    // condition of all the clock gates along the way to the primary clock. All
+    // we have to do is to add this clock gate's condition to that list.
+    auto info = getOrCreateClockLowering(ckgOp.getInput());
+    auto ckgEnable = info.clock.materializeValue(ckgOp.getEnable());
+    info.enable =
+        info.clock.getOrCreateAnd(info.enable, ckgEnable, ckgOp.getLoc());
+    gatedClockLowerings.insert({clock, info});
+    return info;
+  }
+
+  // Create the `ClockTreeOp` that corresponds to this ungated clock.
+  auto &slot = clockLowerings[clock];
+  if (!slot) {
+    auto treeOp = builder.create<ClockTreeOp>(clock.getLoc(), clock, Value{});
+    treeOp.getBody().emplaceBlock();
+    slot = std::make_unique<ClockLowering>(clock, treeOp, stats);
+  }
+  return GatedClockLowering{*slot, Value{}};
+}
+
+ClockLowering &ModuleLowering::getOrCreatePassThrough() {
+  auto &slot = clockLowerings[Value{}];
+  if (!slot) {
+    auto treeOp = builder.create<PassThroughOp>(moduleOp.getLoc());
+    treeOp.getBody().emplaceBlock();
+    slot = std::make_unique<ClockLowering>(Value{}, treeOp, stats);
+  }
+  return *slot;
+}
+
+/// Replace all uses of a value with a `StateReadOp` on a state.
+void ModuleLowering::replaceValueWithStateRead(Value value, Value state) {
+  SmallDenseMap<Block *, Value> readsByBlock;
+  for (auto &use : llvm::make_early_inc_range(value.getUses())) {
+    auto *op = use.getOwner();
+    auto &read = readsByBlock[op->getBlock()];
+    if (!read) {
+      OpBuilder builder(op);
+      read = builder.create<StateReadOp>(value.getLoc(), state);
+    } else {
+      // We have no guarantee that we iterate over the uses in the order
+      // they appear in the block. Therefore make sure we move the read op
+      // before the earliest use.
+      auto *readOp = read.getDefiningOp();
+      if (op->isBeforeInBlock(readOp))
+        readOp->moveBefore(op);
```

Comment:
I did a few (imprecise) measurements on compile time (used rocketConfig and largeBoomConfig from chipyard) of the lower state pass, both in current state and with these two lines removed and the insertion point of the builder always at start of block:

| | current [sec] | modified [sec] | speedup |
|---|---|---|---|
| rocket | ~1.8 | ~0.05 | 36x |
| boom | ~`420 | ~0.3 | 1400x |

I think the reason is that `isBeforeInBlock` has to recompute the ordering every iteration such that its runtime is not O(1) on average as mentioned in its documentation, but the full O(n) where n is the number of ops in the block. Since the body of the module is very big (because it contains all the arc calls, mem ops, etc.) we can clearly see the more than quadratic runtime.

Reply:
Fixed this by creating only one read op initially, and then doing a post-pass where I establish an op order separately and then sink all reads into the blocks where they are used. Could also have just created one read for every use, but that seemed a bit excessive.